File: dsa_leetcode_beginner/0139-word-break/0139-word-break.cpp
#include <vector>
#include <string>
#include <unordered_set>

class Solution {
public:
    bool wordBreak(std::string s, std::vector<std::string>& wordDict) {
        std::unordered_set<std::string> wordSet(wordDict.begin(), wordDict.end());
        std::vector<bool> res(s.length() + 1, false);
        res[0] = true;

        for (int i = 0; i <= s.length(); i++) {
            for (const std::string& word : wordSet) {
                if (res[i] && i + word.length() <= s.length() && s.substr(i, word.length()) == word) {
                    res[i + word.length()] = true;
                }
            }
        }

        return res[s.length()];
    }
};

File: dsa_leetcode_beginner/0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordSet = set(wordDict)
        
        res = [False] * (len(s) + 1)
        res[0] = True
        
        # Iterate through all positions in the string
        for i in range(len(s) + 1):
            # Check all words in the wordSet
            for word in wordSet:
                if (res[i] and (i + len(word)) <= len(s) and s[i : i + len(word)] in wordSet):
                    res[i + len(word)] = True
        
        return res[-1]
        

#Question: https://leetcode.com/problems/word-break
#Blog: https://blog.unwiredlearning.com/word-break